### DISCUSSION TITLE
Exposición «Heartbleed»: preguntas y respuestas

### DISCUSSION THREAD

**Discussion by gwolf** [unamfi/sistop-2022-1#159] [Q&A]
Abro este foro como un espacio de discusión para la exposición que prepararon los alumnos Lucero de la Cruz y Ricardo Carrillo, «Heartbleed». El video está disponible en:

https://youtu.be/IN0WnkDctLU

Pueden consultar también el documento de texto en que los alumnos presentan este mismo tema:

https://github.com/unamfi/sistop-2022-1/blob/main/exposiciones/Carrillo_DeLaCruz/Heartbleed.pdf

Además de participar aquí, les pido que llenen la pequña encuesta que está disponible en:

https://www.questionpro.com/t/ARiwuZp3P5

**Comment by JesusESD** [comment]
Hola compañeros, Lucero y Ricardo! El tema me pareció muy interesante, sobre todo porque también trata sobre una vulnerabilidad, al igual que el mío. Mi duda es la siguiente: durante sus conclusiones, mencionan que Linux fue el sistema operativo que se vio más afectado por Heartbleed, ¿podrían explicarme un poco más a profundidad cuál es la razón de eso? Les agradeceré mucho su respuesta :)

**Reply by ricardo-casa** [reply]
Por los mecanismos que utiliza Linux para asignar memoria! 

No es que sean malos, más bien pueden ser diferentes o que no vienen "por defecto". 

Buscar una comparativa de como es que asignar memoria en Windows y Mac fue difícil para ambos. No encontramos algo concreto pero nos damos una idea del por qué. 🧐

Pero en caso de Linux va muy relacionado con la pregunta que hizo Alberto. Y es que para establecer estos mecanismos hay que buscar herramientas o implementarlas como tal. Una de las recomendaciones que les mencionaba era la sobreescritura de datos una vez que dejemos de trabajar con esos por defecto los compiladores Unix me parece que no lo hace, pero es una buena defensa. 

Los "análisis de memoria" como ASan tampoco es por defecto, lo debemos implementar. 

Por defecto cuando trabajamos con memoria en Linux no se hace nada de eso, inclusive recordando un poco las clases de EDA hasta el free lo teníamos que hacer de manera "manual". Espero haber contestado tu pregunta,  pero si tienes más dudas o algo más concreto haznos saber!

**Reply by gwolf** [reply]
Yo agregaría a la respuesta de Ricardo lo siguiente: Este ataque no es al sistema operativo, sino que a una _biblioteca en espacio de usuario_. Sí, es una biblioteca de sistema — y una muy importante, y utilizada por muchísimos programas... 
La pregunta que hace Jesús Eduardo es muy buena: _Heartbleed_ fue un agujero importante y doloroso, pero no habría sido tan importante si hubiera habido más implementaciones de los estándares criptográficos. Y pasa que implementar criptografía es... Algo que, casi por _mantra_, tenemos que dejar que implementen los que sí lo han estudiado.
OpenSSL pasó hace diez años por un momento de tener muy pocos desarrolladores contribuyendo. Varios proyectos _paralelos_ como SSLEAY fueron perdiendo agilidad y se "secaron" años antes. Y justo hacia 2010 comenzaban a aparecer otros proyectos para ese espacio de problemas — pero el tiempo jugó en contra. Demasiados proyectos estaban basados en OpenSSL, y éste proyecto tomó varias decisiones incorrectas que hacían su código muy difícil de mantener, por lo que no atraía nuevos desarrolladores, formando un círculo vicioso ☹

**Comment by albertogfm** [comment]
Hola compañeros, me pareció muy interesante y útil el tema, mi duda es: cuando hablaron de que para evitar que vuelva a pasar esta vulnerabilidad se deben establecer defensas al momento de asignar memoria, ¿qué medidas exactamente involucra estas defensas ?

**Reply by ricardo-casa** [reply]
En gran parte podemos ver al ataque como un Buffer Overflow, pero no solo vemos una sobre escritura sinó también una sobre lectura (que es el principal problema).

Esa parte que mencionas fue el principal reto que nos encontramos al quererlo sintetizar y consideramos que faltó mucho. Retomando tu pregunta. Los fuzzers son en gran medida una gran herramienta de seguridad para casos de Buffer Overflow pero para heartbleed hace falta un poco más. 

En el artículo que publicó Wheeler una herramienta que actúa específicamente en memoria que sería de gran utilidad para eventos de sobre escritura y sobre lectura; ysok los "desinfectantes de direcciones" (Adress Sanitizer, ASan)

Particularmente esa herramienta se lanzó en el 2012 un año tarde jajaja. Pero es bastante útil inclusive hasta para detectar perdidas de memoria! 

La "desventaja" que encontramos con ASan es que está integrado en los compiladores (GCC y otros), por lo que será necesario volver a compilar todo el código para añadirlo y en algunos casos no se puede :/

Pero es bastante útil y simple de utilizar ya que funciona a base de banderas. 


Si te interesa un poco más acerca de eso puedes checar el github oficial, ya que se trata igualmente de un proyecto open source. ;) 

https://github.com/google/sanitizers/wiki/AddressSanitizer

**Reply by gwolf** [reply]
¡Excelente cómo lo caracterizas como un buffer overflow de sobre-lectura!

**Comment by CarlosGMendozaB** [comment]
Hola compañeros, me pareció muy buena su exposición e interesante. Mi pregunta que tengo es sobre lo que mencionan de las formas de reducir el impacto, ¿Cómo un usuario normal que no tiene conocimiento de este problema, puede identificar este problema o que medidas de prevención puede tomar?

**Reply by ricardo-casa** [reply]
Consideraré a un usuario normal como al usuario final. 

Es bastante simple, sinceramente una que considero apropiada es conocer como es que funcionan los navegadores en sus aspectos más básicos. Podremos notar que algunas veces los certificados están vencidos en algunas de nuestras páginas favoritas y lo ignoramos por completo. Muchas veces será necesario ingresar a sitios así, en ser ese caso, lo mejor sería evitar ingresar datos personales al sitio. 

Cambiar contraseñas constantemente y por nada en el mundo utilizar la misma contraseña para otros sitios. 

La seguridad por parte del usuario final es de lo más básico pero en caso de que el sitio sufra un ataque, con esas medidas nos veremos menos afectados.

**Reply by gwolf** [reply]
Claro, el ataque que presentaron afecta a los servidores, no a los usuarios finales... Excepto en su interacción con servidores 😉
Las recomendaciones de Ricardon son válidas, independientemente de que haya Heartbleed o no.

**Comment by jorgelmp** [comment]
Hola, compañeros. Mi pregunta es: ¿por qué mencionan el nombre de Robin Seegelman en particual? ¿No hubiera sido también la resposabilidad del equipo que mantenían Open SSL? ¿O solo había una persona manteniendo un módulo tan importante del internet?

**Reply by ricardo-casa** [reply]
Pars nada, y estás en todo lo correcto. Erróneamente le atribuimos damos a entender que Seggelmann ha sido el único responsable del error pero no es así y es como tú lo mencionas una serie de responsables. 

Un dato extra que por falta de tiempo puede responder también tu pregunta. Seggelmann era estudiante de doctorado en ese entonces. Su director fue el que autorizo esa versión del protocolo, junto a otros miembros. 

Personalmente considero que como proyecto open source que es. Las empresas o público que ocupa el protocolo también tienen cierta responsabilidad ya que el código es público, sería bueno revisar la documentación antes de instalar protocolos únicamente por qué es una alternativa "gratuita".

**Comment by Inakyord** [comment]
Hola, mi pregunta es: ¿Por qué se tardó tanto tiempo en descubrirse, al menos para los defensores, que existía ésta vulnerabilidad? Además, ¿Existe un antes y después de Heartbleed con respecto a la ciberseguridad, o se tomó como un problema aislado?

**Reply by gwolf** [reply]
Encontrar vulnerabilidades es una tarea complicada, y muchas veces depende de suerte. Suerte de estar suficientemente despierto al leer el código de terceros, suerte de estar buscando el fallo correcto, etc.
2014 fue un año con varios agujeros importantes sobre las validaciones criptográficas en SSL/TLS. Te pongo como ejemplo el siguiente código, de la implementación (independiente) de Apple. ¿Qué error le ves?

    if ((err = SSLHashSHA1.update(&hashCtx, &clientRandom)) != 0)
        goto fail;
    if ((err = SSLHashSHA1.update(&hashCtx, &serverRandom)) != 0)
        goto fail;
    if ((err = SSLHashSHA1.update(&hashCtx, &signedParams)) != 0)
        goto fail;
        goto fail;
    if ((err = SSLHashSHA1.final(&hashCtx, &hashOut)) != 0)
        goto fail;

Claro, te lo facilité mucho diciendo que es una vulnerabilidad obvia y dándote sólo diez líneas, no decenas de miles. Pero el desarrollador de Apple que hizo este código metió un `goto fail` de más. Esto llevó a que las validaciones de seguridad de Apple se fueran al hoyo, y se permitiera la conexión a servidores hostiles (que no tenían el certificado correcto).
https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/
https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2014-1266

**Comment by RGJG28** [comment]
Hola
Me parece interesante el tema, sin embargo creo que entra el factor de seguridad para el usuario final en cuanto a paginas web que no cuentan con el certificado SSL, pero esta certificación tiene un costo anual por lo que algunos sitios no siempre la tienen . ¿saben que riesgos se podria tener al entrar en estos sitios que no cuentan con dicha certificación?

**Reply by gwolf** [reply]
Afortunadamente, (y respondiendo también a @Inakyord), una de las respuestas más importantes a esta vulnerabilidad fue atraer más conciencia a la importancia de la infraestructura criptográfica. Lo que dices era cierto en 2014, pero hoy es posible obtenerlos de forma gratuita, particularmente mediante la autoridad certificadora [Let's Encrypt](https://letsencrypt.org/).
Y la respuesta que pides... Es amplia 😉 Dejémoslo en que, sin un certificado TLS, no vas a poder establecer conexiones cifradas en las que tus usuarios puedan confiar.
